parser/parser.go

```diff
@@ -33,173 +33,151 @@ var (
+// scanUntilFailure scans the output until a failure indicator (--- FAIL:) is
+// found. Returns true if a failure line is detected, false otherwise.
+func (fp *FuzzOutputProcessor) scanUntilFailure(scanner *bufio.Scanner) bool {
+	for scanner.Scan() {
+		line := scanner.Text()
+		fp.logger.Info("Fuzzer output", "message", line)
 
-	// If a failure has not yet been detected, check if this line indicates
-	// a failure.
-	if !fp.State.SeenFailure {
-		if err := fp.handleFailureDetection(line); err != nil {
-			return fmt.Errorf("failure detection failed: %w", err)
+		// Detect the start of a failure section.
+		if strings.Contains(line, "--- FAIL:") {
+			return true
 		}
 	}
+	return false
+}
 
-	// If a failure has been detected, process the line to extract and
-	// record relevant failure information.
-	if fp.State.SeenFailure {
-		// Handle the line in the context of a detected failure.
-		if err := fp.handleFailureLine(line); err != nil {
-			return fmt.Errorf("failure line handling failed: %w",
-				err)
-		}
+// processFailureLines processes lines after a failure is detected, writes them
+// to a log file, and attempts to extract and log the failing input data.
+func (fp *FuzzOutputProcessor) processFailureLines(scanner *bufio.Scanner) {
+	// Construct the log file path for storing failure details.
+	logFileName := fmt.Sprintf("%s_failure.log", fp.targetName)
+	logPath := filepath.Join(fp.cfg.FuzzResultsPath, logFileName)
+
+	// Ensure the results directory exists.
+	if err := utils.EnsureDirExists(fp.cfg.FuzzResultsPath); err != nil {
+		fp.logger.Error("Failed to create fuzz results directory",
+			"error", err)
+		return
 	}
 
-	return nil
-}
+	// Create the log file for writing.
+	logFile, err := os.Create(logPath)
+	if err != nil {
+		fp.logger.Error("Failed to create log file", "error", err)
+		return
+	}
+	fp.logFile = logFile
 
-// handleFailureDetection looks for the first "--- FAIL:" marker. When found,
-// it initializes the failure log file for subsequent lines.
-func (fp *FuzzProcessor) handleFailureDetection(line string) error {
-	// Check if the line contains the failure marker.
-	if strings.Contains(line, "--- FAIL:") {
-		// Mark that a failure has been detected.
-		fp.State.SeenFailure = true
-
-		// Construct the log file name and path for storing failure
-		// details.
-		logFileName := fmt.Sprintf("%s_failure.log", fp.target)
-		logPath := filepath.Join(fp.cfg.FuzzResultsPath, logFileName)
-
-		// Ensure the FuzzResultsPath directory exists (creates parents
-		// as needed)
-		if err := utils.EnsureDirExists(
-			fp.cfg.FuzzResultsPath); err != nil {
-			return fmt.Errorf("failed to create fuzz result file: "+
-				"%w", err)
+	// Ensure the log file is closed at the end.
+	defer func() {
+		if err := fp.logFile.Close(); err != nil {
+			fp.logger.Error("Failed to close log file", "error",
+				err)
 		}
+	}()
+
+	fp.logger.Info("Failure log initialized", "path", logPath)
```

Comment:
Nit:  I think this working is clearer

```suggestion
	fp.logger.Info("Opened failure log for writing", "path", logPath)
```